File: src/git_commit_ai/git.py
"""Git operations for git-commit-ai."""

import subprocess
from dataclasses import dataclass


@dataclass
class DiffResult:
    """Result of a git diff operation."""

    diff: str
    stats: str
    files: list[str]

    @property
    def is_empty(self) -> bool:
        """Check if the diff is empty."""
        return not self.diff.strip()


class GitError(Exception):
    """Exception raised for git operation errors."""

    pass


def run_git(*args: str, check: bool = True) -> str:
    """Run a git command and return the output.

    Args:
        *args: Git command arguments.
        check: Whether to raise an exception on non-zero exit code.

    Returns:
        The command output as a string.

    Raises:
        GitError: If the command fails and check is True.
    """
    try:
        result = subprocess.run(
            ["git", *args],
            capture_output=True,
            text=True,
            check=check,
        )
        return result.stdout.strip()
    except subprocess.CalledProcessError as e:
        raise GitError(f"Git command failed: {e.stderr.strip()}") from e


def get_staged_diff() -> DiffResult:
    """Get the diff of staged changes.

    Returns:
        DiffResult containing the diff, stats, and file list.
    """
    diff = run_git("diff", "--cached")
    stats = run_git("diff", "--cached", "--stat")
    files_output = run_git("diff", "--cached", "--name-only")
    files = [f for f in files_output.split("\n") if f]

    return DiffResult(diff=diff, stats=stats, files=files)


def get_file_diff(file_path: str) -> DiffResult:
    """Get the diff of a specific staged file.

    Args:
        file_path: Path to the file.

    Returns:
        DiffResult for the specific file.
    """
    diff = run_git("diff", "--cached", "--", file_path)
    stats = run_git("diff", "--cached", "--stat", "--", file_path)
    files = [file_path] if diff else []

    return DiffResult(diff=diff, stats=stats, files=files)


def add_files(*paths: str) -> None:
    """Stage files for commit.

    Args:
        *paths: File paths to stage. Use "." for all files.
    """
    if not paths:
        paths = (".",)
    run_git("add", *paths)


def commit(message: str) -> str:
    """Create a commit with the given message.

    Args:
        message: The commit message.

    Returns:
        The commit hash.
    """
    run_git("commit", "-m", message)
    return run_git("rev-parse", "HEAD")


def push() -> None:
    """Push commits to the remote repository."""
    branch = get_current_branch()
    run_git("push", "origin", branch)


def get_current_branch() -> str:
    """Get the name of the current branch.

    Returns:
        The current branch name.
    """
    return run_git("rev-parse", "--abbrev-ref", "HEAD")


def get_modified_files() -> list[str]:
    """Get list of all modified files (staged, unstaged, and untracked).

    Returns:
        List of file paths.
    """
    files = set()

    # Modified and staged files
    staged = run_git("diff", "--cached", "--name-only", check=False)
    if staged:
        files.update(staged.split("\n"))

    # Modified but not staged
    modified = run_git("diff", "--name-only", check=False)
    if modified:
        files.update(modified.split("\n"))

    # Untracked files
    untracked = run_git("ls-files", "--others", "--exclude-standard", check=False)
    if untracked:
        files.update(untracked.split("\n"))

    return [f for f in files if f]


def has_staged_changes() -> bool:
    """Check if there are any staged changes.

    Returns:
        True if there are staged changes, False otherwise.
    """
    diff = run_git("diff", "--cached", "--name-only", check=False)
    return bool(diff.strip())

File: src/git.ts
import { execSync } from "node:child_process";
import type { DiffResult } from "./types.js";

function matchesPattern(filePath: string, pattern: string): boolean {
  // Convert glob pattern to regex
  // Support: *, **, ?
  let regexStr = pattern
    .replace(/\./g, "\\.")
    .replace(/\*\*/g, "<<GLOBSTAR>>")
    .replace(/\*/g, "[^/]*")
    .replace(/<<GLOBSTAR>>/g, ".*")
    .replace(/\?/g, ".");

  // If pattern doesn't start with / or *, it can match anywhere in path
  if (!pattern.startsWith("/") && !pattern.startsWith("*")) {
    regexStr = `(^|/)${regexStr}`;
  }

  // Match end of string or after /
  regexStr = `${regexStr}($|/)`;

  try {
    const regex = new RegExp(regexStr);
    return regex.test(filePath);
  } catch {
    return false;
  }
}

export function shouldIgnoreFile(filePath: string, patterns: string[]): boolean {
  return patterns.some((pattern) => matchesPattern(filePath, pattern));
}

export function filterDiffByPatterns(diff: string, patterns: string[]): string {
  if (!patterns || patterns.length === 0) {
    return diff;
  }

  // Split diff into file sections
  const sections = diff.split(/(?=^diff --git)/m);
  const filtered = sections.filter((section) => {
    // Extract file path from diff header
    const match = section.match(/^diff --git a\/(.+?) b\//m);
    if (!match) {
      return true; // Keep sections without proper header
    }
    const filePath = match[1];
    return !shouldIgnoreFile(filePath, patterns);
  });

  return filtered.join("");
}

export class GitError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "GitError";
  }
}

function runGit(...args: string[]): string {
  try {
    const result = execSync(["git", ...args].join(" "), {
      encoding: "utf-8",
      stdio: ["pipe", "pipe", "pipe"],
    });
    return result.trim();
  } catch (error) {
    const err = error as { stderr?: string; message: string };
    const message = err.stderr?.trim() || err.message;
    throw new GitError(`Git command failed: ${message}`);
  }
}

function runGitSafe(...args: string[]): string {
  try {
    return runGit(...args);
  } catch {
    return "";
  }
}

export function getStagedDiff(): DiffResult {
  const diff = runGitSafe("diff", "--cached");
  const stats = runGitSafe("diff", "--cached", "--stat");
  const filesOutput = runGitSafe("diff", "--cached", "--name-only");
  const files = filesOutput.split("\n").filter((f) => f);

  return {
    diff,
    stats,
    files,
    isEmpty: !diff.trim(),
  };
}

export function getFileDiff(filePath: string): DiffResult {
  const diff = runGitSafe("diff", "--cached", "--", filePath);
  const stats = runGitSafe("diff", "--cached", "--stat", "--", filePath);
  const files = diff ? [filePath] : [];

  return {
    diff,
    stats,
    files,
    isEmpty: !diff.trim(),
  };
}

export function addFiles(...paths: string[]): boolean {
  if (paths.length === 0) {
    paths = ["."];
  }
  try {
    runGit("add", ...paths);
    return true;
  } catch (error) {
    const err = error as GitError;
    // Ignore "ignored file" errors and "pathspec did not match" (deleted files)
    if (
      err.message.includes("ignored by one of your .gitignore") ||
      err.message.includes("pathspec") && err.message.includes("did not match")
    ) {
      return false;
    }
    throw error;
  }
}

export function commit(message: string): string {
  runGit("commit", "-m", `"${message.replace(/"/g, '\\"')}"`);
  return runGit("rev-parse", "HEAD");
}

export function push(): void {
  const branch = getCurrentBranch();
  runGit("push", "origin", branch);
}

export function getCurrentBranch(): string {
  return runGit("rev-parse", "--abbrev-ref", "HEAD");
}

export function getModifiedFiles(): string[] {
  const files = new Set<string>();

  // Modified and staged files
  const staged = runGitSafe("diff", "--cached", "--name-only");
  if (staged) {
    staged.split("\n").forEach((f) => files.add(f));
  }

  // Modified but not staged
  const modified = runGitSafe("diff", "--name-only");
  if (modified) {
    modified.split("\n").forEach((f) => files.add(f));
  }

  // Untracked files
  const untracked = runGitSafe("ls-files", "--others", "--exclude-standard");
  if (untracked) {
    untracked.split("\n").forEach((f) => files.add(f));
  }

  return Array.from(files).filter((f) => f);
}

export function hasStagedChanges(): boolean {
  const diff = runGitSafe("diff", "--cached", "--name-only");
  return Boolean(diff.trim());
}

export function getStagedFiles(): string[] {
  const output = runGitSafe("diff", "--cached", "--name-only");
  return output.split("\n").filter((f) => f);
}

export function resetStaged(): void {
  runGitSafe("reset", "HEAD");
}

export function getLastCommitDiff(): DiffResult {
  const diff = runGitSafe("diff", "HEAD~1", "HEAD");
  const stats = runGitSafe("diff", "HEAD~1", "HEAD", "--stat");
  const filesOutput = runGitSafe("diff", "HEAD~1", "HEAD", "--name-only");
  const files = filesOutput.split("\n").filter((f) => f);

  return {
    diff,
    stats,
    files,
    isEmpty: !diff.trim(),
  };
}

export function commitAmend(message: string): string {
  runGit("commit", "--amend", "-m", `"${message.replace(/"/g, '\\"')}"`);
  return runGit("rev-parse", "HEAD");
}
